smithy-codegen-core/src/main/java/software/amazon/smithy/codegen/core/TraceFile.java

```diff
@@ -0,0 +1,390 @@
+        ObjectNode.Builder shapesBuilder = ObjectNode.objectNodeBuilder();
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            shapesBuilder.withMember(entry.getKey().toString(),
+                    entry.getValue().stream().map(ShapeLink::toNode).collect(ArrayNode.collect()));
+        }
+
+        //returning ObjectNode for TraceFile
+        return ObjectNode.objectNodeBuilder()
+                .withMember(SMITHY_TRACE_TEXT, smithyTrace)
+                .withMember(ARTIFACT_TEXT, artifactMetadata)
+                .withOptionalMember(DEFINITIONS_TEXT, getArtifactDefinitions())
+                .withMember(SHAPES_TEXT, shapesBuilder.build())
+                .build();
+    }
+
+    /**
+     * Finds invalid types and tags and either removes them or throws an error depending on
+     * whether toThrow is true or false.
+     *
+     * @throws ExpectationNotMetException if a type or tag in shapes is not in definitions.
+     */
+    public void validateTypesAndTags() {
+        //The optional ArtifactDefinitions must be non-null to call this method
+        Objects.requireNonNull(artifactDefinitions);
+
+        //for each entry in the shapes map
+        for (Map.Entry<ShapeId, List<ShapeLink>> entry : shapes.entrySet()) {
+            //for each ShapeLink in entry's List<ShapeLink>
+            for (ShapeLink link : entry.getValue()) {
+                //checking if link's type is in artifactDefinitions
+                if (!artifactDefinitions.getTypes().containsKey(link.getType())) {
+                    throw new ExpectationNotMetException(entry.getKey().toString()
+                            + " contains types that aren't in definitions.", sl);
+                }
+
+                //checking if link's tags are all in artifactDefinitions
+                Optional<List<String>> tags = link.getTags();
+                if (tags.isPresent()) {
+                    for (String tag : tags.get())
+                        if (!artifactDefinitions.getTags().containsKey(tag)) {
+                            throw new ExpectationNotMetException(entry.getKey().toString() + " " + tag
+                                    + " is a tag that isn't in definitions.", sl);
+                        }
+                }
+            }
+        }
+    }
+
+    /**
+     * Parses model and determines whether the trace file object meets the specs of the model by checking if
+     * the trace file contains all the ShapeIds in the model and the model contains all the ShapeIDs in the trace file.
+     *
+     * @param modelResourceName the model name to validate the trace file against. Model should be in the
+     *                          resources file.
+     * @throws ExpectationNotMetException if model contains a ShapeID not in TraceFile or TraceFile contains a ShapeID
+     *                                    not in model.
+     */
+    public void validateModel(String modelResourceName) {
```

Comment:
This is another context where a similar separation of concerns comes in to play. This function would need to bloat significantly to handle all the other things that may need to happen to a `ModelAssembler` to build a valid `Model`. Instead, you can have this function just take an instance of an already built `Model`.